### PR TITLE
code

### PR DESCRIPTION
**Pull request by kingmushroom**

### CODE REVIEW

**Review by bitmonkey — CHANGES_REQUESTED** [review]
This is purely a style / code quality review from me:

Flake8 linting brings up lots of things to fix, this is based on me using a 90 char max line length, I've attached the lint results and the contents of the `.flake8` config file I used to set the line length for the linter. 

[enrichment_lint_results.txt](https://github.com/ONSdigital/es-enrichment-sg/files/3410155/enrichment_lint_results.txt)


[flake8.txt](https://github.com/ONSdigital/es-enrichment-sg/files/3410301/flake8.txt)


You can run the linter by doing `./do.sh run python flake8 > lint_results.txt`

I'm looking to build  status  checks in our CI pipeline in future, so when you raise a PR our pipeline will do a build where linting and other status checks are done, if they don't pass we'll have to fix the issues before we can merge.


Other than that just a few tweaks, I'd like us to handle things like potential missing dictionary keys  better e.g looking at the way we're pulling settings from the environment:

instead of:

 ```
my_setting = os.environ['my_setting']
``` 
which will break if that setting hasn't been set it's better to do something like:
```
my_setting = os.environ.get('my_setting', None)
if my_setting is None:
   raise ValueError('my_setting config parameter missing')
```
This will give us an explicit message rather than relying on the catchall error handling and having to work out what's broken from the stack trace.

Building on this approach we can make it even better and use a serialisation framework to validate our expected inputs, there's a library called `marshmallow` we can use for this - see https://pypi.org/project/marshmallow/ +  https://marshmallow.readthedocs.io/en/2.x-line -  which Jamie and David have already started using for their system-of-record API work.

The idea is to make a schema class which models your expected inputs then you can just use that to validate your inputs, it will tell us which fields failed validation and we can just return that in our error response then.

e.g

```
import marshmallow

class InputSchema(marshmallow.Schema):
    county_lookup_column1 = fields.Str(required=True)
    bucket_name = fields.Str(required=True)

def lambda_handler(event, context):
    config, errors =  InputSchema().load(event)
    if errors:
       raise ValueError(f"Error validating environment params: {errors}")
```

The `f"{some_variable}"` pattern above is called an f-string and is a really nice string interpolation template feature in python 3.7 and up.

Final thing it's nicer to handle all the environment vars in one place, in the `lamdba_handler` function, then pass through the values to the individual functions that use them as a function parameter


e.g instead of 

```
def lambda_handler(event, context):
   return some_func()

def somefunc():
   bob = os.environ('some_value')
```

You'd have:

```
def lambda_handler(event, context):
   bob = os.environ.get('some_value', None)
   if bob is None:
      raise ValueError("missing config param bob")
   return some_func(some_value)

def somefunc(some_value):
   bob = os.environ('some_value')
```

Any questions give me a shout, there's already a layer with `marshmalloẁ` in you can pull, and you can add it to your pipfile locally to develop with it.

To help fix the linting issues, you can probably use black e.g

`./do.sh run python black --line-length 90`

That will reformat your python files, the only thing it won't really deal with is comments, they aren't code so it won't be able to build a syntax tree and work out how to change it.

Once it's done can just do `git diff` to see the changes, some things are opinionated, like using double rather than single quotes for strings....but you get used to it and it does save time / mental effort particularly when reading other people's code in the long run as the styles are consistent.

One final thing is to put the data files for your tests in a `fixtures` sub-directory, just makes things easier to find then as you can easily find tests and the data (test fixtures) you are using and its not jumbled up together.

e.g

```
tests/test_enrichment.py
tests/fixtures/test_data.json
```

**Comment by kingmushroom** [comment]
I've pep8'd the code based on those docs. I've aimed to get lines down to 90ish. A handful of lines are above 90 because splitting them to that extent was a bit impractical(happy to go in and alter if you wish, but for example 91 on enrichment method, it doesnt look easy to split it much more)

The environment variables bit seems like its going to be quite a large task(cause it'll affect everything). Happy to do it as part of the review, but it seems big enough to warrant a task of its own.

**Comment by kingmushroom** [comment]
Ah, and i didnt see the fixtures bit till now, i'll do that next

File: enrichment_method.py [hunk 1]
```diff
@@ -17,6 +17,11 @@ def _get_traceback(exception):
         )
     )
 
+def get_environment_variable(variable):
```
Comment (bitmonkey) [inline enrichment_method.py:20]: This is a good first effort  - we've talked about before we should `marshmallow` all the things though.

**Comment by bitmonkey** [comment]
@kingmushroom If we can change the input validation to use `marshmallow` like the other stuff that would be grand. I'll re-review then :)

**Comment by kingmushroom** [comment]
All changes made now I think.

Only thing to note is that if checkpoint isn't supplied it'll fall into a heap. But we discussed this.

**Pull request closed without merging**